index.js

```diff
@@ -20,18 +20,28 @@ export default async function pMap(
 		const result = [];
 		const errors = [];
 		const skippedIndexes = [];
-		const iterator = iterable[Symbol.iterator]();
 		let isRejected = false;
 		let isIterableDone = false;
 		let resolvingCount = 0;
 		let currentIndex = 0;
+		let asyncIterator = false;
+		let iterator;
+
+		if (iterable[Symbol.iterator] === undefined) {
+			// We've got an async iterable
+			iterator = iterable[Symbol.asyncIterator]();
+			asyncIterator = true;
+		} else {
+			iterator = iterable[Symbol.iterator]();
+		}
 
-		const next = () => {
+		const next = async () => {
 			if (isRejected) {
 				return;
 			}
 
-			const nextItem = iterator.next();
+			const nextItem = asyncIterator ? await iterator.next() : iterator.next();
```

Comment:
```suggestion
			const nextItem = await iterator.next();
```

`await` can be applied to synchronous methods too.

Reply:
My understanding on this was that await will work on sync functions but it *will* push them onto the event loop... which introduces stack wrapping for callers that were not even using async and a delay because it's not a direct function call but instead bounces through the event loop.

I didn't write a performance test to confirm the impact of awaiting a sync function... should we do that stand-alone test before deciding which way to go?

Reply:
It will only be pushed until the next micro-task tick. And I don't think it will matter as everything else with `p-map` is async.

Reply:
Hmm... await'ing the non-async `.next()` call causes an existing sync iterator test to fail:

```
test('pMapSkip', async t => {
	t.deepEqual(await pMap([
		1,
		pMapSkip,
		2
	], async value => value), [1, 2]);
});
```

That test receives `[1]` instead of the expected `[1, 2]` if a non-async next() call is awaited.

You can try it yourself by using the line you proposed: `const nextItem = await iterator.next();`.

The behavior change is because `resolve()` on the `pMap()` promise gets called twice and the results change because the `pMapSkip` code has already been called and the 2nd call removes the real second item.

I think you said elsewhere that calling `reject()` twice, likely with two different exceptions, didn't matter... but in this case calling `resolve()` with two different values causes the second value to be returned, which does change the behavior and is an indication of why you should not resolve or reject a promise twice, particularly if you can't guarantee that the values are unchanged.

I'm not entirely sure what to do here.  The expedient solution would be to project the resolve and reject calls.  I think that resolve() getting called twice is *possibly* something that happens in the main branch too, but the lack of awaits causes the behavior change to be hidden unless the caller introduces an event loop bounce before awaiting the returned promise?  I'd like to check that...

Reply:
That test is incorrect and should use `{concurrency: 1}` as otherwise the order cannot be guaranteed. Using `await` here is actually beneficial and no one should be relying on the order of execution when concurrency is not limited to 1.

Reply:
I think the test is actually correct?

Before async iterators and `pMap.next` being async there could only be one outstanding call to `iterator.next()` at a time, but now that async is supported there can be multiple `iterator.next()` calls outstanding and all of them will return done as true if the iterator determines that there are no more items left.  The existing return logic could only get invoked once due to there only being a single pending `iterator.next()` call.  Async caused a problem here because the skip return logic is destructive so it needs protection to ensure that the skip return logic gets invoked only a single time.

I added the skip return logic that is needed in the async version.  Now both the test and the implementation are correct, I believe.

Reply:
I believe this is resolved now as the skip logic will not run twice now.